indexing-service/src/main/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java

```diff
@@ -578,33 +441,158 @@ private void publishSegments(
         Collections.singletonList(sequenceName)
     );
 
-    // Use a separate future to ensure that the publish future is not completed until after
-    // the handoff future is registered in the pending list
-    SettableFuture<SegmentsAndMetadata> publishResultFuture = SettableFuture.create();
+    ListenableFuture<SegmentsAndMetadata> handoffFuture = Futures.transform(publishFuture, driver::registerHandoff);
+
+    pendingHandoffs.add(handoffFuture);
+  }
+
+  private void waitForSegmentHandoff(long handoffTimeout) throws InterruptedException, ExecutionException,
+                                                                 TimeoutException
+  {
+    if (!pendingHandoffs.isEmpty()) {
+      ListenableFuture<?> allHandoffs = Futures.allAsList(pendingHandoffs);
+      log.info("Waiting for handoffs");
+
+
+      if (handoffTimeout > 0) {
+        allHandoffs.get(handoffTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Here, futures are actually the tasks publishing as well as handing off segments. So, the timeout should be `publishAndHandoffTimeout` rather than `handoffTimeout`. I suggest to add a new tuningConfig type for adding a new configuration which is only used in this task.